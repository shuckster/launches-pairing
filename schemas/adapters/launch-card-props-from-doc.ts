import type { TPartialLaunchesDocumentSchema } from '../api/spacex.types';
import type { TLaunchCardSchema } from '../ui/launch-card.types';

/**
 * input is expected to be TLaunchesDocumentSchema, but since
 * it comes from an external API, we can't guarantee that.
 *
 * Could also use Zod et al for schema validation, but trying
 * to keep it basic for this task.
 */

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function launchCardPropsFromApiDoc(apiDoc: any): TLaunchCardSchema {
  const input = (apiDoc || {}) as TPartialLaunchesDocumentSchema;
  return {
    id: input.id ?? null,
    name: input.name ?? null,
    dateUtc: input.date_utc ?? null,
    coreSerial: input.cores?.[0]?.core?.serial ?? null,
    payloadType: input.payloads?.[0]?.type ?? null,
    patchUrl: input.links?.patch?.small ?? null,
    success: input.success ?? null,
    failureReason: input.failures?.[0]?.reason ?? null,
  };
}
