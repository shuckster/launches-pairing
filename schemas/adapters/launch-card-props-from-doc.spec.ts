import { launchCardPropsFromApiDoc } from '@/schemas/adapters/launch-card-props-from-doc';

import mockData from '@/pages/api/launches/mock.json';

/**
 * Happy path
 */

test('launchCardPropsFromApiDoc() snapshot test against mock-data', () => {
  const cards = mockData.docs.map(launchCardPropsFromApiDoc);
  expect(cards).toMatchSnapshot();
});

/**
 * Unhappy path
 */

const empty = {
  id: null,
  name: null,
  dateUtc: null,
  coreSerial: null,
  payloadType: null,
  patchUrl: null,
  success: null,
  failureReason: null,
};

test.each([
  [NaN, empty],
  [undefined, empty],
  [null, empty],
  [0, empty],
  ['', empty],
  [{}, empty],

  // Missing core
  [{ cores: [{ core: { serial: undefined } }] }, empty],
  [{ cores: [{ core: undefined }] }, empty],
  [{ cores: [undefined] }, empty],
  [{ cores: [] }, empty],
  [{ cores: undefined }, empty],

  // Missing payload
  [{ payloads: [{ type: undefined }] }, empty],
  [{ payloads: [undefined] }, empty],
  [{ payloads: [] }, empty],
  [{ payloads: undefined }, empty],

  // Missing links
  [{ links: [{ patch: { small: undefined } }] }, empty],
  [{ links: [{ patch: undefined }] }, empty],
  [{ links: [undefined] }, empty],
  [{ links: [] }, empty],
  [{ links: undefined }, empty],
])(`launchCardPropsFromApiDoc(%p) fails gracefully`, (input, expected) => {
  expect(launchCardPropsFromApiDoc(input)).toEqual(expected);
});
