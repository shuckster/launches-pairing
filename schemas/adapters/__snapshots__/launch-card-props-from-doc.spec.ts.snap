// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`launchCardPropsFromApiDoc() snapshot test against mock-data 1`] = `
[
  {
    "coreSerial": "Merlin1A",
    "dateUtc": "2006-03-24T22:30:00.000Z",
    "failureReason": "merlin engine failure",
    "id": "5eb87cd9ffd86e000604b32a",
    "name": "FalconSat",
    "patchUrl": "https://images2.imgbox.com/94/f2/NN6Ph45r_o.png",
    "payloadType": "Satellite",
    "success": false,
  },
  {
    "coreSerial": "Merlin2A",
    "dateUtc": "2007-03-21T01:10:00.000Z",
    "failureReason": "harmonic oscillation leading to premature engine shutdown",
    "id": "5eb87cdaffd86e000604b32b",
    "name": "DemoSat",
    "patchUrl": "https://images2.imgbox.com/f9/4a/ZboXReNb_o.png",
    "payloadType": "Satellite",
    "success": false,
  },
  {
    "coreSerial": "Merlin1C",
    "dateUtc": "2008-08-03T03:34:00.000Z",
    "failureReason": "residual stage-1 thrust led to collision between stage 1 and stage 2",
    "id": "5eb87cdbffd86e000604b32c",
    "name": "Trailblazer",
    "patchUrl": "https://images2.imgbox.com/6c/cb/na1tzhHs_o.png",
    "payloadType": "Satellite",
    "success": false,
  },
  {
    "coreSerial": "Merlin2C",
    "dateUtc": "2008-09-28T23:15:00.000Z",
    "failureReason": null,
    "id": "5eb87cdbffd86e000604b32d",
    "name": "RatSat",
    "patchUrl": "https://images2.imgbox.com/95/39/sRqN7rsv_o.png",
    "payloadType": "Satellite",
    "success": true,
  },
  {
    "coreSerial": "Merlin3C",
    "dateUtc": "2009-07-13T03:35:00.000Z",
    "failureReason": null,
    "id": "5eb87cdcffd86e000604b32e",
    "name": "RazakSat",
    "patchUrl": "https://images2.imgbox.com/ab/5a/Pequxd5d_o.png",
    "payloadType": "Satellite",
    "success": true,
  },
  {
    "coreSerial": "B0003",
    "dateUtc": "2010-06-04T18:45:00.000Z",
    "failureReason": null,
    "id": "5eb87cddffd86e000604b32f",
    "name": "Falcon 9 Test Flight",
    "patchUrl": "https://images2.imgbox.com/73/7f/u7BKqv2C_o.png",
    "payloadType": "Dragon Boilerplate",
    "success": true,
  },
  {
    "coreSerial": "B0004",
    "dateUtc": "2010-12-08T15:43:00.000Z",
    "failureReason": null,
    "id": "5eb87cdeffd86e000604b330",
    "name": "COTS 1",
    "patchUrl": "https://images2.imgbox.com/fa/dc/FOUDQ0Sn_o.png",
    "payloadType": "Dragon 1.0",
    "success": true,
  },
  {
    "coreSerial": "B0005",
    "dateUtc": "2012-05-22T07:44:00.000Z",
    "failureReason": null,
    "id": "5eb87cdfffd86e000604b331",
    "name": "COTS 2",
    "patchUrl": "https://images2.imgbox.com/c5/f4/XfLVgbaO_o.png",
    "payloadType": "Dragon 1.0",
    "success": true,
  },
  {
    "coreSerial": "B0006",
    "dateUtc": "2012-10-08T00:35:00.000Z",
    "failureReason": null,
    "id": "5eb87ce0ffd86e000604b332",
    "name": "CRS-1",
    "patchUrl": "https://images2.imgbox.com/3e/91/hlGiK49a_o.png",
    "payloadType": "Dragon 1.0",
    "success": true,
  },
  {
    "coreSerial": "B0007",
    "dateUtc": "2013-03-01T19:10:00.000Z",
    "failureReason": null,
    "id": "5eb87ce1ffd86e000604b333",
    "name": "CRS-2",
    "patchUrl": "https://images2.imgbox.com/bd/fe/lXUYKL28_o.png",
    "payloadType": "Dragon 1.0",
    "success": true,
  },
]
`;
